# Exercise 4.9 Intro to Data Viz with Python

This script contains

1. Import libraries
2. Clean customers data
3. Combine cleaned customers data with ords_prods_merged
4. Export to pkl


# 01. Import Libraries

In [1]:
# import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Path to IC folders
path = r"C:\Users\alabaro\OneDrive - DAI\Documents\Aug_2023 Instacart Basket Analysis"

In [3]:
# importing orders_products_merged_v4 from Prepared Data folder
df_ords_prods_merged_v4 = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_v4.pkl'))

In [4]:
# importing customers data from Original Data folder
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

# 02. Clean data

In [5]:
# check output of customers

customers.head(15)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559

In [6]:
customers.shape

(206209, 10)

In [7]:
# review data types
customers.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

Based on output, make the following changes to make all labels consistent
Change First Name to first_name
Change Surnam to last_name
Change STATE to state
Change Gender to sex
Change Age to age

In [8]:
# Rename First Name to first_name

customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [9]:
# Rename Surnam to last_name

customers.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [10]:
# Rename STATE to state

customers.rename(columns = {'STATE' : 'state'}, inplace = True)

In [11]:
# Rename Gender to sex

customers.rename(columns = {'Gender' : 'sex'}, inplace = True)

In [12]:
# Rename Age to age

customers.rename(columns = {'Age' : 'age'}, inplace = True)

In [13]:
# check output

customers.head()

user_id first_name last_name     sex       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [14]:
customers.dtypes

user_id          int64
first_name      object
last_name       object
sex             object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

All lables are now in the correct format. 
Next, do consistency checks for missing values, duplicate values, mixed-type data.

In [15]:
# checking for mixed types

for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

first_name


There are mixed data types in the first_name column. This might be becuase there are null values. 

In [16]:
# checking for null values

customers.isnull().sum()

user_id             0
first_name      11259
last_name           0
sex                 0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [17]:
# look for null values

customer_null = customers[customers['first_name'].isnull() == True]

In [18]:
customer_null

user_id first_name last_name     sex         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

There are 11,259 missing values in the first_name column. Our analysis will not required full names of customers, especially since we have unique user IDs. Also, from a data security perspective, it's probably best to remove identifying information. Therefore, we will create a version dropping first_name and last_name.

In [19]:
# remove first_name and last_name

customers_clean = customers.drop(['first_name','last_name'],axis=1)

In [20]:
# check output

customers_clean.head()

user_id     sex       state  age date_joined  n_dependants fam_status  \
0    26711  Female    Missouri   48    1/1/2017             3    married   
1    33890  Female  New Mexico   36    1/1/2017             0     single   
2    65803    Male       Idaho   35    1/1/2017             2    married   
3   125935  Female        Iowa   40    1/1/2017             0     single   
4   130797  Female    Maryland   26    1/1/2017             1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

New customers_clean dataset has just user id with no first or last names

In [21]:
# checking for duplicates

customers_clean_dups = customers_clean[customers_clean.duplicated()]

In [22]:
customers_clean_dups

Empty DataFrame
Columns: [user_id, sex, state, age, date_joined, n_dependants, fam_status, income]
Index: []

No duplicates found

# 03. Combine customer with ords_prods data

In [23]:
# check output

customers_clean.head()

user_id     sex       state  age date_joined  n_dependants fam_status  \
0    26711  Female    Missouri   48    1/1/2017             3    married   
1    33890  Female  New Mexico   36    1/1/2017             0     single   
2    65803    Male       Idaho   35    1/1/2017             2    married   
3   125935  Female        Iowa   40    1/1/2017             0     single   
4   130797  Female    Maryland   26    1/1/2017             1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

In [24]:
# check output

df_ords_prods_merged_v4.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... _merge     busiest_day    busiest_days  \
0          0  ...   both  Regularly busy  Regularly busy   
1          1  ...   both  Regularly busy    Slowest days   
2          1  ...   both  Regularly busy    Slowest days   
3          1  ...   both      Least busy    Slowest days   
4          1  ...   both      Least busy    Slowest days   

   busiest_period_of_day max_order  loyalty_flag avg_spending spending_flag  \
0         Average orders        10  New customer     6.367797   Low spender   
1         Average orders        10  New customer     6.367797   Low spender   
2            Most orders        10  New customer     6.367797   Low spender   
3         Average orders        10  New customer     6.367797   Low spender   
4            Most orders        10  New customer     6.367797   Low spender   

   med_frequency         frequency_flag  
0           20.5  Non frequent customer  
1           20.5  Non frequent customer  
2           20.5  Non frequent customer  
3           20.5  Non frequent customer  
4           20.5  Non frequent customer  

[5 rows x 24 columns]

Both customers_clean and df_ords_prods_merged_v4 datasets share the user_id variable. Merge based on this key column.

In [26]:
# Merge customers_clean & df_ords_prods_merged_v4

ords_prods_cust_merged = df_ords_prods_merged_v4.merge(customers_clean, on=['user_id'])

In [27]:
# check output

ords_prods_cust_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... spending_flag  med_frequency         frequency_flag     sex  \
0          0  ...   Low spender           20.5  Non frequent customer  Female   
1          1  ...   Low spender           20.5  Non frequent customer  Female   
2          1  ...   Low spender           20.5  Non frequent customer  Female   
3          1  ...   Low spender           20.5  Non frequent customer  Female   
4          1  ...   Low spender           20.5  Non frequent customer  Female   

     state age date_joined n_dependants  fam_status income  
0  Alabama  31   2/17/2019            3     married  40423  
1  Alabama  31   2/17/2019            3     married  40423  
2  Alabama  31   2/17/2019            3     married  40423  
3  Alabama  31   2/17/2019            3     married  40423  
4  Alabama  31   2/17/2019            3     married  40423  

[5 rows x 31 columns]

customers_clean and ords_prods_merged are now combined into one file based on the user_id key column

# 04. Export to pkl

In [28]:
# export combined file as pkl to prepared data folder

ords_prods_cust_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers_merged.pkl'))